### PR TITLE
Misc updates for cisco router configurations

### PR DESCRIPTION
**Pull request by dspicuzzbbn**
Found on a cisco ASA

---
This change is [Reviewable](https://reviewable.io/reviews/batfish/batfish/466)

### CODE REVIEW

File: projects/batfish/src/main/antlr4/org/batfish/grammar/cisco/Cisco_common.g4 [hunk 1]
```diff
@@ -491,6 +491,11 @@ variable_interface_name
    NAME | NEWLINE | TAG | TRACK | VARIABLE )
 ;
 
+variable_password
+:
+   (~NEWLINE)+
```
[the remark this answers is not in the document]
Comment (virtuald) [reply projects/batfish/src/main/antlr4/org/batfish/grammar/cisco/Cisco_common.g4]: I have no desire to learn the deep complexity of ANTLR... so if it's easy for you to fix, by all means do so. We encountered the 'enable password' line on an ASA, so... as long as that doesn't throw a parsing error as it did previously, that's all I care about. :)

I don't think Batfish currently does anything with the password anyways, so if it swallowed the seed then that's probably "ok", though probably not the Right Thing do it.

[the remark this answers is not in the document]
Comment (arifogel) [reply projects/batfish/src/main/antlr4/org/batfish/grammar/cisco/Cisco_common.g4]: That's fine. I'll just finish this one then.

[the remark this answers is not in the document]
Comment (virtuald) [reply projects/batfish/src/main/antlr4/org/batfish/grammar/cisco/Cisco_common.g4]: 👍

File: projects/batfish/src/main/antlr4/org/batfish/grammar/cisco/Cisco_ignored.g4 [hunk 2]
```diff
@@ -237,6 +238,7 @@ null_block
       | TEMPLATE
       | TERMINAL
       | TIME_RANGE
+      | TIMEOUT
```
Comment (arifogel) [inline projects/batfish/src/main/antlr4/org/batfish/grammar/cisco/Cisco_ignored.g4]: This is a breaking change. It needs to be more granular. In particular, it interferes with `ts_timeout`.

Comment (virtuald) [reply projects/batfish/src/main/antlr4/org/batfish/grammar/cisco/Cisco_ignored.g4]: Sad.. it's true, I hadn't ran the unit tests yet. Alright, I'll fix it on Friday. Here are some of the line(s):

```
timeout xlate 3:00:00
timeout pat-xlate 0:00:30
... there are a bunch of other variations also
```

From ASA command reference: https://www.cisco.com/c/en/us/td/docs/security/asa/asa-command-reference/T-Z/cmdref4/t1.html?dtid=osscdc0002833#pgfId-1649892

**Comment by virtuald** [comment]
btw, If you made your other changes (the encrypted variable stuff), you could just cherry pick the first commit from this and push that and I'll fix the second commit later.

**Comment by arifogel** [comment]
I made PR #481 off your first commit, and added grammar changes for password handling.
When #481 is merged, you should update this PR before asking for another review.

File: test_rigs/unit-tests/configs/cisco_enable [hunk 3]
```diff
@@ -6,5 +6,6 @@ enable secret 5 $1$aaaaaaaaaaaaaaaaaaaaaaaaaaa
 enable super-user-password aaaaaaaaaaaaaaaaaaaaa
 enable telnet authentication
 enable telnet password aaaaaaaaaaaaaaaaaaaaaaaaa
+enable password $sha512$5000$98pqLCa2lpNvN6gnQl3zyQ==$QvdeF/4EqAATT2/e64ujg== woldj4
```
Comment (arifogel) [inline test_rigs/unit-tests/configs/cisco_enable]: What configuration language is this? I cannot enter this on my ASA VM. Also, it looks like it contains an invalid base64 string after the last `$`.

Comment (virtuald) [reply test_rigs/unit-tests/configs/cisco_enable]: The password was from an ASA and 'looks like' that (the important part that breaks parsing are the random `$` symbols and the last part) -- but in reality I just randomly changed various characters so it wouldn't match, so I'm not surprised it's not a valid base64 string.

I didn't create the ASA config, @asydney did, so if the password isn't sensitive I suspect he could share the original string if you need it.

Comment (arifogel) [reply test_rigs/unit-tests/configs/cisco_enable]: No I just made it valid base64 by inserting a character. It's just a hash, so I don't need anything original.

**Comment by arifogel** [comment]
#481 has been merged. You should update this branch before your next review request.

**Comment by dspicuzzbbn** [comment]
Updated, should be good to go now.

**Comment by virtuald** [comment]
Alright. I won't have time to get to it until late next week.

**Comment by dspicuzzbbn** [comment]
Sorry for the delay on this, we got redirected and paused for awhile. I added more to the tests, this should be ready to be merged now.

**Comment by arifogel** [comment]
Hey thanks for the update! I just need a few more unit test lines from you. See inline comments.

---

Reviewed 6 of 6 files at r1.
Review status: all files reviewed at latest revision, 3 unresolved discussions.

---

*[projects/batfish/src/main/antlr4/org/batfish/grammar/cisco/Cisco_ignored.g4, line 919 at r1](https://reviewable.io:443/reviews/batfish/batfish/466#-L0LuwUwNhF6DFjZxpcp:-L0LuwUwNhF6DFjZxpcq:b-6w90ip) ([raw file](https://github.com/batfish/batfish/blob/0a248b485aa1eaafaaf9dc3bea9dbfbb08d23984/projects/batfish/src/main/antlr4/org/batfish/grammar/cisco/Cisco_ignored.g4#L919)):*
> ```ANTLR
>             | HALF_CLOSED
>             | ICMP
>             | ICMP_ERROR
> ```

Missing unit test.

---

*[projects/batfish/src/main/antlr4/org/batfish/grammar/cisco/Cisco_ospf.g4, line 194 at r1](https://reviewable.io:443/reviews/batfish/batfish/466#-L0Luyb9MByNmUsvYXNT:-L0Luyb9MByNmUsvYXNU:b-6w90ip) ([raw file](https://github.com/batfish/batfish/blob/0a248b485aa1eaafaaf9dc3bea9dbfbb08d23984/projects/batfish/src/main/antlr4/org/batfish/grammar/cisco/Cisco_ospf.g4#L194)):*
> ```ANTLR
>       )
>       | LOG
>       | LOG_ADJ_CHANGES
> ```

Missing unit test.

---

*[projects/batfish/src/main/antlr4/org/batfish/grammar/cisco/CiscoLexer.g4, line 10678 at r1](https://reviewable.io:443/reviews/batfish/batfish/466#-L0Lv1wTbA97A4Hlv6fd:-L0Lv1wTbA97A4Hlv6fe:bsn29ns) ([raw file](https://github.com/batfish/batfish/blob/0a248b485aa1eaafaaf9dc3bea9dbfbb08d23984/projects/batfish/src/main/antlr4/org/batfish/grammar/cisco/CiscoLexer.g4#L10678)):*
> ```ANTLR
> ;
> 
> TCP_INSPECTION
> ```

Missing parser rule referencing this token, and missing unit test.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/batfish/batfish/466#-:-L0LrW4vxdGqbE7fep12:b12oktl)*

**Comment by dspicuzzbbn** [comment]
Updated.

**Comment by arifogel** [comment]
:lgtm:

---

Reviewed 3 of 3 files at r2, 1 of 1 files at r3.
Review status: all files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/batfish/batfish/466#-:-L0mtI0lEmlKA7qA_PMq:bnfp4nl)*

**Pull request merged**